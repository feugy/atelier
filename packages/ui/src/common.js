import { locale, translations, getBrowserLocale } from 'svelte-intl'
import './styles.css'
import fr from './locales/fr.yaml'
import en from './locales/fr.yaml'

translations.update({ fr, en })
locale.set(getBrowserLocale('fr'))
