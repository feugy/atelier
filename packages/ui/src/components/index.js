export { default as EventLogger } from './EventLogger.svelte'
export { default as Explorer } from './Explorer.svelte'
export { default as Group } from './Group.svelte'
export { default as PaneContainer } from './PaneContainer.svelte'
export { default as Toolbar } from './Toolbar.svelte'
