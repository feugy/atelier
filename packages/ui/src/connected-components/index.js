export { default as App } from './App.svelte'
export { default as ErrorDialogue } from './ErrorDialogue.svelte'
export { default as EventsPane } from './EventsPane.svelte'
export { default as PropertiesPane } from './PropertiesPane.svelte'
