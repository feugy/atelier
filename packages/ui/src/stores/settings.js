import { BehaviorSubject, map, tap } from 'rxjs'
import Ajv from 'ajv'

const ajv = new Ajv()
const settings$ = new BehaviorSubject()
reloadSettings()

export function getSettings(key, schema = {}) {
  const validate = ajv.compile(schema)

  return settings$.pipe(
    map(settings => settings?.[key]),
    tap(value => {
      if (value !== undefined && !validate(value)) {
        throw new Error(
          `Invalid value for "${key}" UI settings:\n${ajv.errorsText(
            validate.errors
          )}`
        )
      }
    })
  )
}

export function reloadSettings() {
  settings$.next(window.uiSettings)
}
