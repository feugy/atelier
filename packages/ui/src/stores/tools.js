import { BehaviorSubject, map, scan, shareReplay } from 'rxjs'
import { groupByName } from '../utils'

let workframe = null
let workframeOrigin = null
const tools$ = new BehaviorSubject([])
const current$ = new BehaviorSubject()
const events$ = new BehaviorSubject()
const lastError$ = new BehaviorSubject()

function updateUrl(fullName) {
  if (fullName) {
    const url = new URL(window.location)
    url.searchParams.set('tool', fullName)
    window.history.pushState({ fullName }, '', url)
  }
}

function readToolFromUrl() {
  const params = new URLSearchParams(window.location.search)
  return params.get('tool')
}

function handleMessage({ origin, data }) {
  if (origin === workframeOrigin) {
    if (data.type === 'registerTool') {
      registerTool(data.data)
    } else if (data.type === 'recordEvent') {
      events$.next({ ...data, time: Date.now() })
    } else if (data.type === 'recordError') {
      const error = new Error(data.message)
      error.stack = data.stack
      lastError$.next(error)
    }
  }
}

function registerTool(tool) {
  const { value: list } = tools$

  const idx = list.findIndex(({ fullName }) => fullName === tool.fullName)

  tools$.next([
    ...(idx === -1 ? list : [...list.slice(0, idx), ...list.slice(idx + 1)]),
    tool
  ])

  // the current tool has been updated, or there are no current tool, but an url match
  const urlName = readToolFromUrl()
  if (
    (idx >= 0 && current$.value === list[idx]) ||
    (!current$.value && (urlName === tool.fullName || !urlName))
  ) {
    current$.next(tool)
  }
}

function postMessage(type, data) {
  if (workframe) {
    workframe.contentWindow?.postMessage({ type, data }, workframeOrigin)
  }
}

current$.subscribe(data => {
  postMessage('selectTool', data)
  updateUrl(data?.fullName)
})

window.addEventListener('popstate', ({ state }) => {
  if (state?.fullName) {
    for (const tool of tools$.value) {
      if (state.fullName === tool.fullName) {
        current$.next(tool)
        break
      }
    }
  }
})

export const tools = tools$.pipe(map(groupByName))

export const currentTool = current$.asObservable()

export const lastError = lastError$.asObservable()

export const events = events$.pipe(
  // reset log when receiving null
  scan((log, event) => (event ? [event, ...log] : []), []),
  shareReplay(1)
)
// subscribe so we never loose any event
events.subscribe()

export function setWorkbenchFrame(frame) {
  workframe = frame
  workframeOrigin = new URL(workframe.src).origin
  current$.next()
  tools$.next([])
  window.removeEventListener('message', handleMessage)
  window.addEventListener('message', handleMessage)
}

export function selectTool(tool) {
  if (tools$.value.includes(tool)) {
    current$.next(tool)
    clearEvents()
  }
}

export function updateProperty({ detail } = {}) {
  if (current$.value && detail instanceof Object) {
    postMessage('updateProperty', { ...detail, tool: current$.value })
  }
}

export function clearEvents() {
  events$.next(null)
}
