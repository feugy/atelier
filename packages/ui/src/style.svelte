<style global type="postcss">
  @import 'material-design-icons-iconfont';

  @tailwind base;

  .material-icons {
    vertical-align: -0.2em;
  }

  html,
  body {
    @apply font-sans h-full overflow-hidden m-0;
    color: theme('colors.primary.text');
    background-color: theme('background.page');
  }

  body {
    @apply flex;
  }

  button {
    @apply outline-none;
    transition: transform 250ms;

    &:focus,
    &:hover {
      transform: scale(1.2);
    }
  }

  .tool.documented {
    @apply p-4;

    & .tool-preview {
      @apply my-4 bg-white p-2 shadow-md;
    }
  }
</style>
