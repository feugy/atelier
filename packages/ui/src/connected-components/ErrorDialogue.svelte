<script>
  import { _ } from 'svelte-intl'
  import { Dialogue } from '../components'
  import { lastError } from '../stores'
</script>

{#if $lastError}
  <Dialogue title={$_('title.error')}>
    <main>{$lastError.message}</main>
    <div slot="footer" class="text-sm">
      {$_('message.fix-and-reload')}
    </div>
  </Dialogue>
{/if}
