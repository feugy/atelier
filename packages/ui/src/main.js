import { App } from './connected-components'

new App({ target: document.body })
