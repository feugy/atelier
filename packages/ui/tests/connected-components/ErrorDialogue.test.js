import { render, screen, within } from '@testing-library/svelte'
import html from 'svelte-htm'
import { translate } from '../test-utils'
import ErrorDialogue from '../../src/connected-components/ErrorDialogue'
import { lastError } from '../../src/stores'

jest.mock('../../src/stores/tools', () => ({
  lastError: new (require('rxjs').BehaviorSubject)()
}))

describe('ErrorDialogue connected component', () => {
  beforeEach(() => {
    lastError.next()
  })

  it('displays nothing without error', () => {
    render(html`<${ErrorDialogue} />`)
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('displays error message', () => {
    const message = 'some random error text'
    lastError.next(new Error(message))
    const { container } = render(html`<${ErrorDialogue} />`)
    const dialogue = screen.getByRole('dialog')
    expect(dialogue).toBeInTheDocument()
    expect(within(dialogue).getByRole('heading')).toHaveTextContent(
      translate('title.error')
    )
    expect(within(dialogue).getByRole('main')).toHaveTextContent(message)
    expect(within(dialogue).getByRole('contentinfo')).toHaveTextContent(
      translate('fix-and-reload')
    )
    expect(container).toMatchSnapshot()
  })
})
