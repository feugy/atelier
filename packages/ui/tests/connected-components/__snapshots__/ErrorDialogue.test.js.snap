// Vitest Snapshot v1

exports[`ErrorDialogue connected component > displays error message 1`] = `
<body>
  <div>
    <!--&lt;Portal&gt;-->
    <!--&lt;Dialogue&gt;-->
    
    <!--&lt;ErrorDialogue&gt;-->
  </div>
  <div>
    <div
      class="filter s-OcGGhMDsfjw6 open"
    />
     
    <div
      class="backdrop s-OcGGhMDsfjw6 open"
    >
      <article
        class="s-OcGGhMDsfjw6"
        role="dialog"
      >
        <header
          aria-level="1"
          class="s-OcGGhMDsfjw6"
          role="heading"
        >
          title.error
        </header>
         
        <div
          class="content s-OcGGhMDsfjw6"
        >
          <main>
            some random error text
          </main>
        </div>
         
        <footer
          class="s-OcGGhMDsfjw6"
        >
          <div
            class="text-sm"
            slot="footer"
          >
            message.fix-and-reload
          </div>
        </footer>
      </article>
    </div>
  </div>
</body>
`;
